### PR TITLE
SVM: Add specification document

### PR DESCRIPTION
**Pull request by dmakarov**

### CODE REVIEW

File: svm/doc/spec.md [hunk 1]
```diff
@@ -0,0 +1,241 @@
+# Solana Virtual Machine specification
+
+# Introduction
+
+Several components of the Solana Validator are involved in processing
+of a transaction (or a batch of transactions).  Collectively the
```
Comment (LucasSte) [inline svm/doc/spec.md]:
```suggestion
a transaction (or a batch of transactions).  Collectively, the
```

File: svm/doc/spec.md [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+components responsible for transaction execution are designated as
+Solana Virtual Machine (SVM). SVM packaged as a stand-alone library
+can be used in applications outside the Solana Validator.
+
+This document represents specification of the SVM. It covers the API
```
Comment (LucasSte) [inline svm/doc/spec.md]:
```suggestion
This document represents the SVM specification. It covers the API
```

File: svm/doc/spec.md [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+- **Modified SVM (SVM+)**
+
+    An SVM type with all the current functionality and extended
+    instructions for custom use cases. This would form a superset of
+    the current SVM.
+
+# System Context
+
+In this section SVM is represented as a single entity. We describe its
```
Comment (LucasSte) [inline svm/doc/spec.md]:
```suggestion
In this section, SVM is represented as a single entity. We describe its
```

File: svm/doc/spec.md [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+interfaces to the parts of the Solana Validator external to SVM.
+
+In the context of Solana Validator the main entity external to SVM is
```
Comment (LucasSte) [inline svm/doc/spec.md]:
```suggestion
In the context of Solana Validator, the main entity external to SVM is
```

File: svm/doc/spec.md [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+bank. It creates an SVM, submits transactions for execution and
+receives results of transaction execution from SVM.
+
+![context diagram](/svm/doc/diagrams/context.svg "System Context")
+
+## Interfaces
+
+In this section we describe the API of using the SVM both in Solana
```
Comment (LucasSte) [inline svm/doc/spec.md]:
```suggestion
In this section, we describe the API of using the SVM both in Solana
```

File: svm/doc/spec.md [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+Validator and in third-party applications.
+
+The interface to SVM is represented by the
+`transaction_processor::TransactionBatchProcessor` struct.  To create
+a `TransactionBatchProcessor` object the client need to specify the
```
Comment (LucasSte) [inline svm/doc/spec.md:83]:
```suggestion
a `TransactionBatchProcessor` object the client needs to specify the
```

File: svm/doc/spec.md [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+`slot`, `epoch`, `epoch_schedule`, `fee_structure`, `runtime_config`,
+and `program_cache`.
+
+The main entry point to the SVM is the method
+`load_and_execute_sanitized_transactions`. In addition
```
Comment (LucasSte) [inline svm/doc/spec.md:88]:
```suggestion
`load_and_execute_sanitized_transactions`. In addition,
```

File: svm/doc/spec.md [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    - a vector of `TransactionLoadResult`, and
+    - a vector of `TransactionExecutionResult`.
+
+An integration test `svm_integration` contains an example of
+instantiating `TransactionBatchProcessor` and calling its method
+`load_and_execute_sanitized_transactions`.
+
+# Functional Model
+
+In this section we describe the functionality (logic) of the SVM in
```
Comment (LucasSte) [inline svm/doc/spec.md]:
```suggestion
In this section, we describe the functionality (logic) of the SVM in
```

File: svm/doc/spec.md [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+terms of its components, relationships among components, and their
+interactions.
+
+On a high level the control flow of SVM consists of loading program
+accounts, checking and verifying the loaded accounts, creating
+invocation context and invoking RBPF on programs implementing the
+instructions of a transaction. The SVM needs to have access to account
+database, sysvar cache via traits implemented for the corresponding
```
Comment (LucasSte) [inline svm/doc/spec.md]:
```suggestion
instructions of a transaction. The SVM needs to have access to an account
database, and a sysvar cache via traits implemented for the corresponding
```

File: svm/doc/spec.md [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+            - Return `struct LoadedTransaction` containing the accounts (pubkey and data),
+              indices to the excutabe accounts in `TransactionContext` (or `InstructionContext`),
+              the transaction rent, and the `struct RentDebit`.
+            - Generate a `NonceFull` struct (holds fee subtracted nonce info) when possible, `None` otherwise.
+    - Returns `TransactionLoadedResult`, a tuple containing the `LoadTransaction` we obtained from `loaded_transaction_accounts`,
+      and a `Option<NonceFull>`.
+
+3. Execute each loaded transactions
+   1. Compute the sum of transaction account balances. This sum is
+      invariant of the transaction execution.
```
Comment (LucasSte) [inline svm/doc/spec.md]:
```suggestion
   1. Compute the sum of transaction accounts' balances. This sum is
      invariant in the transaction execution.
```

File: svm/doc/spec.md [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+      - If the account `RentState` pre-transaction processing is rent exempt or unitiliazed, the verification will pass.
+      - If the account `RentState` pre-transaction is rent paying:
+         - A transition to a state uninitialized or rent exempt post-transaction is not allowed.
+         - If its size has changed or its balance has increased, it cannot remain rent paying.
+   8. Extract log messages.
+   9. Extract inner instructions (`Vec<Vec<InnerInstruction>>`).
+   10. Extract `ExecutionRecord` components from transaction context.
+   11. Check balances of accounts to match the sum of balances before
+       transaction execution.
+   12. Updated loaded transaction accounts to new accounts.
```
Comment (LucasSte) [inline svm/doc/spec.md]:
```suggestion
   12. Update loaded transaction accounts to new accounts.
```

File: svm/doc/spec.md [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+   13. Extract changes in accounts data sizes
+   14. Extract return data
+   15. Return `TransactionExecutionResult` with wrapping the extracted
+       information in `TransactionExecutionDetails`.
+
+4. Prepare the results of loading and executing transactions.
+
+   This includes the following steps for each transactions
+   1. Dump flattened result to info log for an account which pubkey is
```
Comment (LucasSte) [inline svm/doc/spec.md]:
```suggestion
   1. Dump flattened result to info log for an account whose pubkey is
```

**Review by LucasSte — COMMENTED** [review]
LGTM. I only pointed small nitpicks.

**Comment by dmakarov** [comment]
> LGTM. I only pointed small nitpicks.

Thank you. I committed the suggested changes.

**Review by LucasSte — APPROVED** [review]

**Pull request merged**